File: solutions/python/flower-field/1/flower_field.py
def annotate(garden):
    if not garden:
        return []
    
    # Validate all rows have same length
    if len(garden) > 0:
        first_len = len(garden[0])
        for row in garden:
            if len(row) != first_len:
                raise ValueError("The board is invalid with current input.")
    
    # Validate all characters are valid
    for row in garden:
        for char in row:
            if char not in ' *':
                raise ValueError("The board is invalid with current input.")
    
    rows = len(garden)
    cols = len(garden[0]) if rows > 0 else 0
    if cols == 0:
        return [""] * rows
        
    result = []
    for r in range(rows):
        row = ""
        for c in range(cols):
            if garden[r][c] == '*':
                row += '*'
            else:
                count = 0
                # Check 8 neighbors
                for dr in [-1, 0, 1]:
                    for dc in [-1, 0, 1]:
                        if dr == 0 and dc == 0:
                            continue
                        nr, nc = r + dr, c + dc
                        if 0 <= nr < rows and 0 <= nc < cols and garden[nr][nc] == '*':
                            count += 1
                row += str(count) if count > 0 else ' '
        result.append(row)
    return result

File: solutions/javascript/flower-field/1/flower-field.js
export const annotate = (input) => {
  const rows = input.length;
  if (rows === 0) return [];
  const cols = input[0].length;
  const result = input.map((row) => row.split(''));

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (result[r][c] === '*') continue;
      let count = 0;
      for (let dr = -1; dr <= 1; dr++) {
        for (let dc = -1; dc <= 1; dc++) {
          if (dr === 0 && dc === 0) continue;
          const nr = r + dr;
          const nc = c + dc;
          if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && input[nr][nc] === '*') {
            count++;
          }
        }
      }
      result[r][c] = count === 0 ? ' ' : String(count);
    }
  }

  return result.map((row) => row.join(''));
};
